projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/base/LLVMTerminatorNode.java

```diff
@@ -29,25 +29,46 @@
  */
 package com.oracle.truffle.llvm.nodes.impl.base;
 
-import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
 import com.oracle.truffle.api.frame.VirtualFrame;
-import com.oracle.truffle.llvm.nodes.base.LLVMNode;
+import com.oracle.truffle.api.instrumentation.Instrumentable;
+import com.oracle.truffle.api.source.SourceSection;
+import com.oracle.truffle.llvm.nodes.base.LLVMStatementNode;
 
 /**
  * This node represents a terminator instruction in LLVM IR. This node decides which basic block is
  * executed next, or terminates execution of a function.
  *
+ * Since this class has an additional field, it can't be simply wrapped by the wrapper node
+ * generated from super class {@link LLVMStatementNode}. Therefore this class is also {@link Instrumentable}.
+ *
  * @see <a href="http://llvm.org/docs/LangRef.html#terminator-instructions">terminator
  *      instructions</a>
  */
-public abstract class LLVMTerminatorNode extends LLVMNode {
+@Instrumentable(factory = LLVMTerminatorNodeWrapper.class)
+public abstract class LLVMTerminatorNode extends LLVMStatementNode {
 
-    @CompilationFinal private final int[] successors;
```

Comment:
The Java `final` just specifies that the reference is final. You can still modify the array content, e.g., do something like `successors[3] = 0;`. The `@CompilationFinal` annotation tells the compiler that also the array content is final and is thus not redundant.

It should not be a problem to not have the source section in `LLVMStatementNode` `@CompilationFinal` since accesses to it are not on the fast path anyway.